### PR TITLE
Migration and adaptation of tier 1 manager, agent and WPK package generation tools

### PR DESCRIPTION
**Pull request by jotacarma90**
|Related issue|
|---|
|https://github.com/wazuh/internal-devel-requests/issues/712|

## Description

In this PR, we will include all the necessary changes to perform the complete migration of the tier 1 package generation tools.

All these changes are reflected in the following issues, and their respective PRs:
- [x] https://github.com/wazuh/wazuh/issues/22320
- [x] https://github.com/wazuh/wazuh/issues/22080
- [x] https://github.com/wazuh/wazuh/issues/22868
- [x] https://github.com/wazuh/wazuh/issues/22839
- [x] https://github.com/wazuh/wazuh/issues/22321
- [x] https://github.com/wazuh/wazuh/issues/22367
- [x] https://github.com/wazuh/wazuh-agent-packages/issues/17

## Tests

Issue with complete testing related to all workflows package generation: 
- https://github.com/wazuh/wazuh-agent-packages/issues/21

### CODE REVIEW

**Review by lchico — CHANGES_REQUESTED** [review]
Some comments.

File: packages/README.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    "ref":"22080-migrate-and-adapt-the-wazuh-agent-subsystem",
+    "inputs":
+        {
+         "docker_image_tag":"auto",
+         "architecture":"amd64",
+         "system":"deb",
+         "revision":"test",
+         "is_stage":"false",
+         "legacy":"false",
+         "checksum":"false",
```
[the remark this answers is not in the document]
Comment (MarcelKemp) [reply packages/README.md:120]: The ID is only used in the workflow to distinguish them, so there is no point in adding it in the README.

File: .github/workflows/packages-upload-images.yml [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+          GH_TOKEN: ${{ secrets.CI_WAZUH_AGENT_PACKAGES }}
+
+      - name: Request pkg_rpm_agent_builder_armhf update
+        if: steps.changes.outputs.pkg_rpm_agent_builder_armhf == 'true'
+        run: |
+          gh workflow run packages-upload-agent-images-arm.yml --repo wazuh/wazuh-agent-packages -r ${{ github.ref_name }} -f docker_image_tag=${{ env.TAG }} -f system=rpm -f architecture=armhf -f source_reference=${{ github.ref_name }}
+        env:
+          GH_TOKEN: ${{ secrets.CI_WAZUH_AGENT_PACKAGES }}
+
+      - name: Request compile_windows_agent update
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/workflows/packages-upload-images.yml:192]: `linux_wpk_builder_x86_64` removed

File: .github/workflows/packages-upload-images.yml [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+              - 'packages/build.sh'
+              - 'packages/generate_package.sh'
+              - 'packages/rpms/arm64/agent/**'
+              - 'packages/rpms/utils/**'
+            pkg_rpm_agent_builder_armhf:
+              - 'packages/build.sh'
+              - 'packages/generate_package.sh'
+              - 'packages/rpms/armhf/agent/**'
+              - 'packages/rpms/utils/**'       
+            compile_windows_agent:
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/workflows/packages-upload-images.yml:90]: `linux_wpk_builder_x86_64` removed

File: .github/actions/ghcr-pull-and-push/retag_image.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+retag_image(){
+    DOCKER_IMAGE_NAME="$1"
+    OLD_TAG="$2"
+    NEW_TAG="$3"
+    GITHUB_REPOSITORY="wazuh/wazuh"
+    GITHUB_OWNER="wazuh"
+    IMAGE_ID=ghcr.io/${GITHUB_OWNER}/${DOCKER_IMAGE_NAME}
+    IMAGE_ID=$(echo ${IMAGE_ID} | tr '[A-Z]' '[a-z]')
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/actions/ghcr-pull-and-push/retag_image.sh:35]: done

File: .github/workflows/packages-filebeat.yml [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        with:
+          aws-access-key-id: ${{ secrets.CI_INTERNAL_DEVELOPMENT_BUCKET_USER_ACCESS_KEY }}
+          aws-secret-access-key: ${{ secrets.CI_INTERNAL_DEVELOPMENT_BUCKET_USER_SECRET_KEY }}
+          aws-region: ${{ secrets.CI_AWS_REGION }}
+
+      - name: Package files, change permissions, create tar and upload to S3
+        working-directory: extensions/filebeat/7.x/wazuh-module
+        run: |
+          mkdir wazuh
+          cp -r _meta wazuh/; cp -r alerts wazuh/; cp -r archives wazuh/; cp -r module.yml wazuh/
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/workflows/packages-filebeat.yml:58]: done

File: .github/actions/ghcr-pull-and-push/retag_image.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    # Bring old tag
+    pull_output=$(docker pull ${IMAGE_ID}:${OLD_TAG})
+
+    if echo "$pull_output" | grep -qi "error"; then
+        echo "Failed pulling ${IMAGE_ID}:${OLD_TAG}"
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/actions/ghcr-pull-and-push/retag_image.sh:41]: done

File: packages/README.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+```bash
+curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $GH_WORKFLOW_TOKEN" -H "X-GitHub-Api-Version: 2022-11-28" --data-binary "@$(pwd)/wazuh-agent-test-amd64-rpm.json" "https://api.github.com/repos/wazuh/wazuh/actions/workflows/packages-upload-agent-images.yml/dispatches"
+```
+
+Where the JSON looks like this:
+
+```json
+# cat wazuh-agent-test-amd64-rpm.json
+{
+    "ref":"22080-migrate-and-adapt-the-wazuh-agent-subsystem",
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/README.md:88]: done

File: packages/README.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    * amd64 -> x86_64
+    * arm64 -> aarch64
+    * armhf -> armv7hl
+
+# Workflow
+
+## Generate and push builder images to GH
+
+```bash
+curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $GH_WORKFLOW_TOKEN" -H "X-GitHub-Api-Version: 2022-11-28" --data-binary "@$(pwd)/wazuh-agent-test-amd64-rpm.json" "https://api.github.com/repos/wazuh/wazuh/actions/workflows/packages-upload-agent-images.yml/dispatches"
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/README.md:80]: it should be `"https://api.github.com/repos/wazuh/wazuh/actions/workflows/packages-upload-agent-images-amd.yml/dispatches"`

File: packages/README.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+         "is_stage":"false",
+         "legacy":"false"
+        }
+}
+```
+
+## Generate packages
+
+```json
+curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $GH_WORKFLOW_TOKEN" -H "X-GitHub-Api-Version: 2022-11-28" --data-binary "@$(pwd)/wazuh-agent-test-amd64-rpm.json" "https://api.github.com/repos/wazuh/wazuh/actions/workflows/packages-build-linux-agent.yml/dispatches"
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/README.md:104]: done

File: packages/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/README.md:120]: not needed, it is not related to the script functioning

File: packages/macos/generate_wazuh_packages.sh [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+    # Define output package name
+    if [ $IS_STAGE == "no" ]; then
+        pkg_name="wazuh-agent_${VERSION}-${REVISION}_${ARCH}_${short_commit_hash}"
+    else
+        pkg_name="wazuh-agent-${VERSION}-${REVISION}.${ARCH}"
+    fi
+
+    get_pkgproj_specs $VERSION $pkg_name
+
+
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/macos/generate_wazuh_packages.sh:155]: done

File: packages/macos/generate_wazuh_packages.sh [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+        fi
+        clean_and_exit 0
+    else
+        echo "ERROR: something went wrong while building the package."
+        clean_and_exit 1
+    fi
+}
+
+function help() {
+
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/macos/generate_wazuh_packages.sh:185]: done

File: packages/macos/generate_wazuh_packages.sh [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+        echo "Error: architecture not supported."
+        echo "Supported architectures: intel64, arm64"
+        exit 1
+    fi
+
+    if [[ "${BUILD}" != "no" ]]; then
+        check_root
+        build_package
+        "${CURRENT_PATH}/uninstall.sh"
+    fi
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/macos/generate_wazuh_packages.sh:434]: added newlines to separate if blocks

File: packages/macos/package_files/build.sh [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    cp ${SOURCES_PATH}/etc/templates/config/generic/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/generic/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/15/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/15/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/16/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/16/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/17/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/17/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/18/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/18/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/19/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/19/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/20/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/20/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/21/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/21/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/22/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/22/
+    cp ${SOURCES_PATH}/etc/templates/config/darwin/23/sca.files ${INSTALLATION_SCRIPTS_DIR}/sca/darwin/23/
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply packages/macos/package_files/build.sh:78]: done

**Review by lchico — APPROVED** [review]
LGTM!

**Review by jotacarma90 — COMMENTED** [review]
LGTM

File: packages/generate_package.sh [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+        REVISION="${REVISION}.el5"
+        TAR_URL="https://packages-dev.wazuh.com/utils/centos-5-i386-build/centos-5-i386.tar.gz"
+        TAR_FILE="${CURRENT_PATH}/${SYSTEM}s/${ARCHITECTURE}/legacy/centos-5-i386.tar.gz"
+        if [ ! -f "$TAR_FILE" ]; then
+            download_file ${TAR_URL} "${CURRENT_PATH}/${SYSTEM}s/${ARCHITECTURE}/legacy"
+        fi
+        DOCKERFILE_PATH="${CURRENT_PATH}/${SYSTEM}s/${ARCHITECTURE}/legacy"
+        # TODO: Remove the "pkg" appended to the container name.
+        # This was necessary because we don't have permission to overwrite an existing container.
+        # Also, ensure that the same adjustment is made in the else condition.
```
Comment (MarcelKemp) [inline packages/generate_package.sh:68]: Remove this TO DO, as it is not a problem.

File: packages/wpk/README.md
Comment (MarcelKemp) [inline packages/wpk/README.md:1]: Update this README with the refactor made to the WPKs.

**Review by MarcelKemp — CHANGES_REQUESTED** [review]

File: packages/README.md [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+**Example Usage:**
+
+1. Build a manager package for amd64 architecture:
+./wazuh_package_builder.sh -t manager -a amd64 -s /tmp --system rpm
+
+2. Build a debug agent package for i386 architecture with checksum generation:
+./wazuh_package_builder.sh -t agent -a i386 -s /tmp -d -c --system rpm
+
+3. Build a legacy RPM package for CentOS 5 (manager):
+./wazuh_package_builder.sh -t manager -l -s /tmp --system rpm
```
Comment (MarcelKemp) [inline packages/README.md:62]: Generate for a legacy agent.

File: packages/generate_package.sh (hunk 14 above)
Comment (jotacarma90) [reply packages/generate_package.sh]: done

File: packages/wpk/README.md
Comment (jotacarma90) [reply packages/wpk/README.md:1]: done

File: packages/README.md (hunk 15 above)
Comment (jotacarma90) [reply packages/README.md]: done

**Review by mjcr99 — APPROVED** [review]
LGTM

**Review by MarcelKemp — APPROVED** [review]
LGTM. GJ team!

**Pull request merged**